C-Plus-Plus/ds/avl_tree.cpp

```diff
@@ -0,0 +1,481 @@
+#include <iostream>
+#include <bits/stdc++.h>
+using namespace std;
+struct Node{
+    Node* left;
+    Node* right;
+    int data;
+    int bf;//balance factor of the node
+    int height;
+};
+
+int find_height(Node* current_node){
+    if(current_node==NULL) return(0);
+    //cur_node is the pointer to the current Node
+    Node* node=current_node;
+    //The balance factor >0 means that left subtree has greater height
+    //if it is <0 then, right subtree has greater height
+    if(node->bf>=0){
+        node = node->left;
+    }
+    else if(node->bf<0){
+        node = node->right;
+    }
+    int height=0;
+    while(node!=NULL){
+        height++;
+        if(node->bf<0){
+            node = node->right;
+        }
+        else if(node->bf>=0){
+            node = node->left;
+        }
+    }
+    //one is added as the height of a single node is 1 and is added to
+    //the height of the longest subtree.
+    return(1+height);
+}
+
+//current node here is the critical node which has a balance factor either
+// greater than  1 or less than -1.
+
+int find_balance_factor(Node* node){
+    return(find_height(node->left)-find_height(node->right));
+}
+
+Node* right_rotate(Node* critical_node){//right rotation
+    Node* next = critical_node->left;
+    Node* temp = next->right;
+    next->right=critical_node;
+    critical_node->left=temp;
+
+    //updating height and balance factors for the rotated nodes
+    next->height=find_height(next);
+    next->bf = find_balance_factor(next);
+
+    critical_node->height=find_height(critical_node);
```

Comment:
The code lacks comments. Please add comments to it. Comments should above the code segment and not alongside.